ruby/red-arrow/lib/arrow/slicer.rb

```diff
@@ -351,5 +356,33 @@ def evaluate
         BooleanArray.new(raw_array)
       end
     end
+
+    class MatchSubstringCondition < Condition
+      def initialize(function, column, pattern, ignore_case, invert: false)
+        @function = function
+        @column = column
+        @options = MatchSubstringOptions.new
+        @options.pattern = pattern
+        @options.ignore_case = ignore_case
+        @invert = invert
+      end
+
+      def !@
+        MatchSubstringCondition.new(@function,
+                                    @column,
+                                    @options.pattern,
+                                    @options.ignore_case?,
+                                    invert: !@invert)
+      end
+
+      def evaluate
+        datum = Function.find(@function).execute([@column.data], @options)
+        if @invert
+          Function.find("invert").execute([datum]).value
+        else
+          datum.value
+        end
```

Comment:
It's not a strong opinion but the following may be simpler:

```suggestion
        if @invert
          datum = Function.find("invert").execute([datum])
        end
        datum.value
```